packages/flutter/lib/src/widgets/pages.dart

```diff
@@ -105,5 +115,108 @@ class PageRouteBuilder<T> extends PageRoute<T> {
   Widget buildTransitions(BuildContext context, Animation<double> animation, Animation<double> secondaryAnimation, Widget child) {
     return transitionsBuilder(context, animation, secondaryAnimation, child);
   }
+}
+
+/// A page that creates a [PageRoute] with customizable transition.
+///
+/// Similar to the [PageRouteBuilder], callers must define the [pageBuilder]
+/// function which creates the route's primary contents. To add transitions
+/// define the [transitionsBuilder] function.
+///
+/// See also:
+///
+///  * [PageRouteBuilder], which is a [PageRoute] version of this class.
+class TransitionPage<T> extends Page<T> {
+  /// Creates a [TransitionPage].
+  const TransitionPage({
+    @required this.pageBuilder,
+    this.transitionsBuilder = _defaultTransitionsBuilder,
+    this.transitionDuration = const Duration(milliseconds: 300),
+    this.opaque = true,
+    this.barrierDismissible = false,
+    this.barrierColor,
+    this.barrierLabel,
+    this.maintainState = true,
+    this.fullscreenDialog = false,
+    LocalKey key,
+    String name,
+    Object arguments,
+  }) : assert(pageBuilder != null),
+      assert(transitionsBuilder != null),
+      assert(opaque != null),
+      assert(barrierDismissible != null),
+      assert(maintainState != null),
+      assert(fullscreenDialog != null),
+      super(key: key, name: name, arguments: arguments);
+
+  /// {@macro flutter.widgets.pageRouteBuilder.pageBuilder}
+  final RoutePageBuilder pageBuilder;
+
+  /// {@macro flutter.widgets.pageRouteBuilder.transitionsBuilder}
+  final RouteTransitionsBuilder transitionsBuilder;
+
+  /// {@macro flutter.widgets.transitionRoute.transitionDuration}
+  final Duration transitionDuration;
+
+  /// {@macro flutter.widgets.transitionRoute.opaque}
+  final bool opaque;
+
+  /// {@macro flutter.widgets.modalRoute.barrierDismissible}
+  final bool barrierDismissible;
+
+  /// {@macro flutter.widgets.modalRoute.barrierColor}
+  final Color barrierColor;
+
+  /// {@macro flutter.widgets.modalRoute.barrierLabel}
+  final String barrierLabel;
+
+  /// {@macro flutter.widgets.modalRoute.maintainState}
+  final bool maintainState;
+
+  /// {@macro flutter.widgets.pageRoute.fullscreenDialog}
+  final bool fullscreenDialog;
+
+  @override
+  Route<T> createRoute(BuildContext context) => _PageBasedPageRouteBuilder<T>(this);
```

Comment:
Why the indirection via _PageBasedPageRouteBuilder? Could it just return a PageRouteBuilder?

Reply:
PageRouteBuilder takes in builder methods and the they cannot be changed through out the route lifecycle. Let's say someone use update the page with same key but different builder, the route underneath will update its settings to be the new page, but it cannot update its builders to be same as the page's builders.

The _PageBasedPageRouteBuilder that builds its transition through the pages builder. So when page updates, the builders it use will be up to date as well. The same philosophy applies to MaterialPage and CupertinoPage

Reply:
Does that actually work, though? From what I can tell, the builder is called from `Route.buildPage`, which is only called when the route first builds (as documented). In fact, I believe its result is actually cached somewhere and reused for all subsequent builds of the route. So even if the builder changes, I would expect it to never take affect.

The tests in this PR seem to indicate that it does take effect, though. How are we clearing the cached page?

Reply:
I didn't see where we cache it in our code, maybe the document is outdated? Even if the caching is true, using _PageBasedPageRouteBuilder will just make the TransitionPage.pageBuilder as responsive as thePageRouteBuilder.pageBuilder but not less.

Reply:
The pages update in navigator will called changedExternalState which clear the cache

Reply:
This is not good, it seems like we only want to call this when pages list actually change. Currently, it is calling whenever the navigator rebuild, and that that is wasteful

Reply:
well, maybe not that wasteful, because it is just widget build and we do that all the time

Reply:
Interesting. I believe there are some other properties that are not honored when they are changed after the route build for the first time. `maintainState` is one example I can think of.

Reply:
I do a little search it looks like `maintainState` is the only persistent flag that doesn't change through out the route lifecycle.

MaterialPage:
builder: ChangedExternalState
fullScreenDialog: ChangedExternalState

CupertinoPage:
builder: ChangedExternalState
fullScreenDialog: ChangedExternalState
title: ChangedExternalState

TransitionPage:
pageBuilder: ChangedExternalState
transitionsBuilder: changedInternalState
transitionDuration: changedInternalState
opaque: changedInternalState
barrierDismissible: changedInternalState
barrierColor: changedInternalState
barrierLabel: changedInternalState
fullscreenDialog:ChangedExternalState

for me, it seems like this is still a reasonable thing to do, what do you think?

Reply:
`OverlayEntry.maintainState` is a setter, could we call it in `changedInternalState` to sync the Route's maintainState with the OverlayEntry again to make the flag take effect?